### PR TITLE
Feature/plugin convex

### LINKED ISSUE #684: Convex 3pool

### LINKED ISSUE #685: Convex stable metapools (with fraxbp + 3pool)
- [ ] MIM + 3CRV
- [ ] ALUSD+FRAXBP
- [ ] LUSD+FRAXBP
- [x] eUSD+FRAXBP

### LINKED ISSUE #722: Convex v2 tricrypto

### LINKED ISSUE #743: Reward token asset deployments + registration (COMP/CRV/CVX/stkAAVE)
- have contracts
- just need to deploy them

### PR DESCRIPTION
**Pull request by tbrent**
Resolves #743 
Resolves #685 
Resolves #684 
Resolves #722

### CODE REVIEW

File: contracts/plugins/assets/convex/PoolTokens.sol [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+            require(_t3timeout1 > 0, "t3timeout1 zero");
+            require(_t3error1 < FIX_ONE, "t3error1 too large");
+        }
+    }
+
+    // === Public Views ===
+
+    /// @return low {UoA/ref_index}
+    /// @return high {UoA/ref_index}
+    function tokenPrice(uint8 index) public view returns (uint192 low, uint192 high) {
```
Comment (pmckelvy1) [inline contracts/plugins/assets/convex/PoolTokens.sol:235]: add doc for param

File: contracts/plugins/assets/convex/PoolTokens.sol [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+        for (uint8 i = 0; i < feeds.length; ++i) {
+            maxLength = uint8(Math.max(maxLength, feeds[i].length));
+        }
+        return maxLength;
+    }
+
+    // x and y can be any two fixes that can be multiplied
+    /// @param xErr {1} error associated with x
+    /// @param yErr {1} error associated with y
+    // returns low and high extremes of x * y, given errors
```
Comment (pmckelvy1) [inline contracts/plugins/assets/convex/PoolTokens.sol:334]: fix up docs

File: contracts/plugins/assets/convex/PoolTokens.sol [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+    function minFeedsLength(AggregatorV3Interface[][] memory feeds) private pure returns (uint8) {
+        uint8 minLength = type(uint8).max;
+        for (uint8 i = 0; i < feeds.length; ++i) {
+            minLength = uint8(Math.min(minLength, feeds[i].length));
+        }
+        return minLength;
+    }
+
+    function maxFeedsLength(AggregatorV3Interface[][] memory feeds) private pure returns (uint8) {
```
Comment (pmckelvy1) [inline contracts/plugins/assets/convex/PoolTokens.sol:323]: docs

File: contracts/plugins/assets/convex/PoolTokens.sol [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+    function getToken(uint8 index) private view returns (IERC20Metadata) {
+        if (index >= nTokens) revert WrongIndex(nTokens - 1);
+        if (index == 0) return token0;
+        if (index == 1) return token1;
+        if (index == 2) return token2;
+        return token3;
+    }
+
+    function minFeedsLength(AggregatorV3Interface[][] memory feeds) private pure returns (uint8) {
```
Comment (pmckelvy1) [inline contracts/plugins/assets/convex/PoolTokens.sol:315]: docs

File: contracts/plugins/assets/convex/PoolTokens.sol [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+            uint192 balance = shiftl_toFix(curvePool.balances(i), -int8(token.decimals()));
+            balances[i] = (balance);
+        }
+
+        return balances;
+    }
+
+    // === Private ===
+
+    function getToken(uint8 index) private view returns (IERC20Metadata) {
```
Comment (pmckelvy1) [inline contracts/plugins/assets/convex/PoolTokens.sol:307]: docs

File: contracts/plugins/assets/convex/CvxStableCollateral.sol [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+        // {UoA}
+        (uint192 aumLow, uint192 aumHigh) = totalBalancesValue();
+
+        // {UoA/tok} = {UoA} / {tok}
+        pegPrice = ((aumLow + aumHigh) / 2).div(supply); // use avg of aumLow + aumHigh
+
+        // discount aumLow by the amount of revenue being hidden
+        // {UoA} = {UoA} * {ref/tok} / {ref/tok}
+        aumLow = aumLow.mulDiv(exposedReferencePrice, _underlyingRefPerTok());
```
Comment (pmckelvy1) [inline contracts/plugins/assets/convex/CvxStableCollateral.sol:73]: why not just `aumLow.mul(revenueShowing)`?

File: contracts/plugins/assets/convex/CvxVolatileCollateral.sol [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    function _anyDepegged() internal view override returns (bool) {
+        uint192[] memory balances = getBalances(); // [{tok}]
+        uint192[] memory vals = new uint192[](balances.length); // {UoA}
+        uint192 valSum; // {UoA}
+
+        // Calculate vals
+        for (uint8 i = 0; i < nTokens; i++) {
+            try this.tokenPrice(i) returns (uint192 low, uint192 high) {
+                // {UoA/tok} = {UoA/tok} + {UoA/tok}
+                uint192 mid = low + high / 2;
```
Comment (pmckelvy1) [inline contracts/plugins/assets/convex/CvxVolatileCollateral.sol:97]: need parens `uint192 mid = (low + high) / 2;`

**Review by pmckelvy1 — CHANGES_REQUESTED** [review]

File: scripts/deployment/phase2-assets/collaterals/deploy_convex_stable_plugin.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      targetName: ethers.utils.formatBytes32String('USD'),
+      priceTimeout: PRICE_TIMEOUT,
+      chainlinkFeed: DAI_USD_FEED, // unused but cannot be zero
+      oracleError: bn('1'), // unused but cannot be zero
+      oracleTimeout: bn('1'), // unused but cannot be zero
+      maxTradeVolume: MAX_TRADE_VOL,
+      defaultThreshold: DEFAULT_THRESHOLD,
+      delayUntilDefault: DELAY_UNTIL_DEFAULT,
+    },
+    bn('1e14'), // revenueHiding = 0.01%
```
Comment (tbrent) [inline scripts/deployment/phase2-assets/collaterals/deploy_convex_stable_plugin.ts:84]: We should either remove revenueHiding entirely or set it to the value in `scripts/deployment/utils.ts` (fp('1e-6')), given that we think the virtual price cannot decrease.

**Comment by tbrent** [comment]
@akshatmittal verification script for 3pool is missing btw

**Comment by pmckelvy1** [comment]
lgtm

**Pull request merged**